src/NServiceBus.Core/Unicast/Messages/MessageMetadataRegistry.cs

```diff
@@ -143,6 +170,7 @@ static IEnumerable<Type> GetParentTypes(Type type)
 
         Conventions conventions;
         Dictionary<RuntimeTypeHandle, MessageMetadata> messages = new Dictionary<RuntimeTypeHandle, MessageMetadata>();
+        List<Tuple<string, Type>> cachedTypes = new List<Tuple<string, Type>>();
```

Comment:
As @timbussmann and I discussed earlier today, my first approach to fixing this perf issue was to change the `messages` key to be a string instead instead of a `RuntimeTypeHandle`, using the `AssemblyQualifedName` as the key value.

This made the string version of ` MessageMetadataRegistry.GetMessageMetadata` run in around 45ms total time, but it ended up slowing down the version of this method that takes a string. It went from 10ms to 150ms.

Turns out that the `Type.AssemblyQualifedName` property is rather expensive, because it creates the string each time you access it! [link](http://referencesource.microsoft.com/#mscorlib/system/rttype.cs,3807)

That led me to the approach of keeping the `RuntimeTypeHandle` key and caching the `GetType` call instead. I was trying to get the performance as close possible to the 45ms of the string approach, so that's why I went with the list of tuples instead of a dictionary. For small numbers of items, iterating a list can be faster than looking up values in a dictionary.

I took a little bit of time today to see where the break even point is, and from what I'm seeing on my machine, once you have more than 5 items in the list, it becomes faster to use the dictionary.

For this collection, it will end up with a number of items equal to the unique message types the endpoint receives, so I'm thinking that it would probably be better to use a dictionary here after all, even if that does mean it will be a bit slower for < 5 message types.